File: classes/Algoritmi/path/2020-04-02/hanoi.cpp
#include<iostream>
#include<cassert>

using namespace std;

unsigned long long int num_mosse(int n) {
// ritorna il minimo numero di mosse che consente di spostare una torre di Hanoi di n dischi
    assert( n >= 0);
    if (n==0)
        return 0;
    return 2*num_mosse(n-1)+1;    
}

void sposta_disco(int n, char piolo_from, char piolo_to) {
    cout << "sposta il disco " << n << " da " << piolo_from << " a " << piolo_to << endl;
}

void sposta_torre(int n, char piolo_from, char piolo_to, char piolo_aux) {
// sposta la torre di Hanoi di n dischi
    assert( n >= 0);
    if (n==0)
        return;
    //int piolo_aux = 6 - piolo_from -piolo_to;
    sposta_torre(n-1, piolo_from, piolo_aux, piolo_to);
    sposta_disco(n, piolo_from, piolo_to);  
    sposta_torre(n-1, piolo_aux, piolo_to, piolo_from);    
}

int main() {
    int N;
    cin >> N;
    cout << "Num mosse: " << num_mosse(N) << endl;
    sposta_torre(N,'A','C','B');
    return 0;
}

File: classes/RO/2019-2020/path/2020-04-02/hanoi.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

def num_mosse_chiusa(N):
    return 2**N-1


def num_mosse(N):
    """Ritorna in minimo numero di mosse necessario per spostare una torre di N dischi."""   
    assert N >= 0
    if N==0:
        return 0
    return 2*num_mosse(N-1) +1

def sposta_disco(N, piolo_from, piolo_to):
    print(f"sposta il disco {N} dal piolo {piolo_from} al piolo {piolo_to}")

def sposta_torre(N, piolo_from, piolo_to, piolo_aux):
    """Sposta una torre di Hanoi di <N> dischi
     dal piolo <piolo_form> al piolo <piolo_to>,
     utilizzando <piolo_aux> come spazio di disgaggio,
     e impiegando il minor numero possibile di mosse."""   
    assert N >= 0
    if N==0:
        return
    sposta_torre(N-1,piolo_from, piolo_aux, piolo_to)
    sposta_disco(N, piolo_from, piolo_to)
    sposta_torre(N-1,piolo_aux, piolo_to, piolo_from)

N = int(input())
assert num_mosse(N) == num_mosse_chiusa(N)
print(f"OK {N}")
#print(f"il numero di mosse che serve per spostare una torre di {N} dischi è {num_mosse(N)} mosse.")
#print("Eccole:")
#sposta_torre(N,'A','C','B')